### DISCUSSION TITLE
Wrong derivation for SH trigonometric terms

### DISCUSSION THREAD

**Discussion by wumo** [google/filament#5845] [Ideas]
As wrote in [9.6.1 Basis functions](https://google.github.io/filament/Filament.html#annex/sphericalharmonics/basisfunctions):

> It’s also fairly easy to compute the trigonometric terms recursively:
> 
> $$ 
> \begin{align*} 
> C_m &\equiv cos(m \phi) \\ 
> S_m &\equiv sin(m \phi) \\ 
> \{ x, y, z \} &= \{ cos \phi sin \theta, sin \phi sin \theta, cos \theta \} 
> \end{align*}
> $$
> 
> Using the angle sum trigonometric identities:
> 
> $$
> \begin{align*}
> cos(m \phi + \phi) &= cos(m \phi) cos(\phi) - sin(m \phi) sin(\phi) \Leftrightarrow C_{m + 1} = \frac{(x C_m - y S_m)}{sin(\theta)^{|m + 1|}} \\
> sin(m \phi + \phi) &= sin(m \phi) sin(\phi) + cos(m \phi) sin(\phi) \Leftrightarrow S_{m + 1} = \frac{(x S_m - y C_m)}{sin(\theta)^{|m + 1|}}
> \end{align*}
> $$

If we define $C_m \equiv cos(m \phi)$ and $S_m \equiv sin(m \phi)$, then the identity formula(which is also typed wrong in the doc):

 $$
sin(m \phi + \phi) = sin(m \phi) cos(\phi) + cos(m \phi) sin(\phi)
$$ 

should be $S_{m+1} = S_{m} cos(\phi) + C_{m} sin(\phi)$, not $S_{m + 1} = \frac{(x S_m - y C_m)}{sin(\theta)^{|m + 1|}}$.

The correct definition for this should be (according to [Appendix A1 in Stupid Spherical Harmonics (SH) Tricks](https://www.ppsloan.org/publications/StupidSH36.pdf)):

$$
\begin{align*}
C_m &\equiv cos(m \phi) (sin\theta)^m\\
S_m &\equiv sin(m \phi) (sin\theta)^m\\
\{ x, y, z \} &= \{ cos \phi sin \theta, sin \phi sin \theta, cos \theta \}
\end{align*}
$$

Then the angle sum trigonometric identities will be:

$$
\begin{align*} 
&sin(m \phi + \phi) = sin(m \phi) cos(\phi) + cos(m \phi) sin(\phi) \\
&\Leftrightarrow \frac{S_{m + 1}}{(sin\theta)^{m+1}} = \frac{S_m}{(sin\theta)^m} cos\phi + \frac{C_m}{(sin\theta)^m} sin\phi \\
&\Leftrightarrow S_{m + 1} = S_m sin\theta cos\phi + C_m sin\theta sin\phi \\
&\Leftrightarrow S_{m + 1} = xS_m + yC_m
\end{align*}
$$

The same derivation can be applied to $cos(m\phi)$:

$$
\begin{align*} 
cos(m \phi + \phi) = cos(m \phi) cos(\phi) - sin(m \phi) sin(\phi) \Leftrightarrow C_{m + 1} = xC_m - yS_m
\end{align*}
$$

The result is:

$$
\begin{align*} 
&cos(m \phi )= \frac{C_m}{(sin\theta)^m} \\
&sin(m \phi ) =  \frac{S_m}{(sin\theta)^m}
\end{align*}
$$

**Comment by romainguy** [comment]
@pixelflinger

**Comment by pixelflinger** [comment]
Thanks, I will review this.

**Comment by pixelflinger** [comment]
Good catch for the typo in the angle sum trigonometric identities.
